service/view_building_coordinator.cc

```diff
@@ -139,6 +147,139 @@ future<std::optional<view_building_coordinator::vbc_state>> view_building_coordi
     co_return state;
 }
 
+static std::optional<dht::token_range> get_range_to_build(const locator::tablet_map& tablet_map, const dht::token_range_vector ranges) {
+    for (auto& range: ranges) {
+        auto tid = tablet_map.get_tablet_id(range.end().value().value());
+        auto trinfo = tablet_map.get_tablet_transition_info(tid);
+
+        if (!trinfo) {
+            return range;
+        }
+    }
+    return std::nullopt;
+}
+
+static std::pair<std::vector<view_name>, dht::token_range> get_views_and_range_for_target(replica::database& db, table_id base_id, const base_tasks& base_tasks, const view_building_target& target) {
+    std::vector<view_name> views;
+    std::optional<dht::token_range> range;
+
+    auto& tmap = db.get_token_metadata().tablets().get_tablet_map(base_id);
+    for (auto& [view, tasks]: base_tasks) {
+        if (!tasks.contains(target)) {
+            continue;
+        }
+
+        if (!range) {
+            auto maybe_range = get_range_to_build(tmap, tasks.at(target));
+            if (maybe_range) {
+                range = std::move(maybe_range);
+                views.push_back(view);
+            }
+        } else {
+            auto& target_tasks = tasks.at(target);
+            if (std::find(target_tasks.cbegin(), target_tasks.cend(), *range) != target_tasks.cend()) {
+                views.push_back(view);
+            }
+        }
+    }
+
+    if (!range) {
+        return {{}, dht::token_range()};
+    }
+    return {std::move(views), *range};
+}
+
+future<> view_building_coordinator::build_view(vbc_state state) {
+    if (!state.currently_processed_base_table) {
+        vbc_logger.info("No view to process");
+        co_return;
+    }
+
+    SCYLLA_ASSERT(state.tasks.contains(*state.currently_processed_base_table));
```

Comment:
I think that `SCYLLA_ASSERT` is a bit too harsh - this will crash the node that runs it. Given that this code runs on the group0 leader, you are risking introducing a situation where a half of the cluster gets killed (or stuck in a restart loop) by a bug:

1. The current group0 leader has a bug and crashes on this assertion
2. If there is at least quorum of nodes, a different node is chosen as the group0 leader, then GOTO 1

It's better to use `on_internal_error` here - this will cause a crash in non-release mode, but in release mode it will only be an exception, so it will be much less painful.